primitives/src/transaction.rs

```diff
@@ -538,7 +538,19 @@ impl Decoder for TransactionDecoder {
             State::Outputs(..) => Err(E(Inner::EarlyEnd("outputs"))),
             State::Witnesses(..) => Err(E(Inner::EarlyEnd("witnesses"))),
             State::LockTime(..) => Err(E(Inner::EarlyEnd("locktime"))),
-            State::Done(tx) => Ok(tx),
+            State::Done(tx) => {
```

Comment:
I think it would be better for this logic to be on Transaction instead of nested in the decoder.  Possibly an invariant that prevents a Transaction from being created that's greater than `MAX_MONEY`.

Reply:
> We are not going to close the Transaction struct and make it only constructible by methods that enforce output amount limits. It's tempting but would be extremely annoying.

At least if Transaction had a method like `valid?()` that returned a bool if it was valid or not, that would be more re-usable than this code on Decoder.  Then decoder (or maybe a future constructor) could just call `.valid?()` or whatever.

Reply:
I would prefer to just discuss decoding here, where we have a fairly-simple criteria "we should pass the Bitcoin Core test vectors as long as they're not obviously crazy".

In general I don't mind rejecting things that are guaranteed to be invalid on-chain. But we cannot reject everything (because we don't have a script interpreter, don't have access to a UTXO set, and we don't know all future softforks, etc etc).

But having an `is_valid` method which checks a grab-bag of arbitrary validity criteria would just be confusing (and then returns `true` if they all pass, I guess?). It's hard for me to imagine why a user would ever want to call such a method.